### DISCUSSION TITLE
failed to connect to all addresses error

### DISCUSSION THREAD

**Discussion by depinglin85** [googleapis/google-cloud-cpp#11964] [Q&A] [answered]
Hello, I would like to develop a tool based on Google's Text-to-Speech. I have encountered some issues. The quickstart example compiles and works correctly, but when integrating it into my project, I encounter the following error:

"Error in non-idempotent operation SynthesizeSpeech: failed to connect to all addresses; last error: UNKNOWN: ipv6:[2404:6800:4004:823::200a]:443: Network is unreachable."

Can you please tell me the reason behind this?

My development environment is Debian.

**Comment by coryan** [comment]
The error message suggests this is a network problem. Can you connect to the endpoint?  Maybe try:

```shell
gcloud info --run-diagnostics
```

Or (even simpler):

```shell
curl -i https://speech.googleapis.com/
```

That second command should return a 404 error, which at least indicates the server was contacted.  If connectivity works: are you using some kind of HTTP proxy?  If so, you may need to configure `grpc_proxy`:

https://grpc.github.io/grpc/cpp/md_doc_environment_variables.html

You may also want to test using the google cloud CLI, e.g.:

https://cloud.google.com/speech-to-text/docs/transcribe-gcloud

**Reply by depinglin85** [reply]
Thank you for your response.
I have figured out the reason for this issue. It is because my OpenSSL version is incompatible with the OpenSSL version required by gRPC.
Can gRPC be used with OpenSSL 1.x?

**Reply by coryan** [reply]
>  I have figured out the reason for this issue. It is because my OpenSSL version is incompatible with the OpenSSL version required by gRPC. 

Okay, that is possible.

> Can gRPC be used with OpenSSL 1.x?

Yes, gRPC can be used with OpenSSL 1.x.  I just tested this (we do test this in our CI builds too):

```
ldd cmake-out/quickstart-speech/quickstart  | grep ssl
        libssl.so.10 => /lib64/libssl.so.10 (0x00007f5d51f30000)
```

```
 env GOOGLE_CLOUD_CPP_USER_PROJECT=${MY_PROJECT} cmake-out/quickstart-speech/quickstart 
results {
  alternatives {
    transcript: "hello"
    confidence: 0.989034295
  }
  result_end_time {
    nanos: 450000000
  }
  language_code: "en-us"
}
total_billed_time {
  seconds: 1
}
request_id: 241832806413851113
```


Whatever the problem is, it must be more involved than just the OpenSSL version.  I do not think we have ruled out networking problems.  It could also be a problem with OpenSSL certificates.  Maybe try to enable logging in the Cloud Speech library:

https://cloud.google.com/cpp/docs/reference/speech/latest/speech-env

and/or in gRPC:

https://github.com/grpc/grpc/blob/master/doc/environment_variables.md

**Reply by depinglin85** [reply]
Thank you very much for your response. 
I also used the ldd command to inspect my quickstart, and I noticed that there is no SSL-related information.
`	linux-vdso.so.1 (0x00007ffed3c8a000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007feba23d1000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007feba2204000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007feba20c0000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007feba20a6000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007feba1ed2000)
	/lib64/ld-linux-x86-64.so.2 (0x00007feba388e000)`

I am using Debian as my operating system and compiling gRPC and Text-to-Speech using the following commands.
`git -C /usr/local/src clone https://github.com/microsoft/vcpkg.git
cd /usr/local/src/vcpkg/
./bootstrap-vcpkg.sh
./vcpkg install google-cloud-cpp[core,speech,texttospeech]
cmake -H. -B.build -DCMAKE_TOOLCHAIN_FILE=/usr/local/src/vcpkg/scripts/buildsystems/vcpkg.cmake
cmake --build .build`

compile the quickstart
`cmake --build .build
  cmake -H. -B.build -DCMAKE_TOOLCHAIN_FILE=/usr/local/src/vcpkg/scripts/buildsystems/vcpkg.cmake`

`./quick/.build/quickstart 
results {
  alternatives {
    transcript: "hello"
    confidence: 0.989034295
  }
  result_end_time {
    nanos: 450000000
  }
  language_code: "en-us"
}
total_billed_time {
  seconds: 1
}
request_id: 3922693910287336567`

Is there anything wrong with the steps I followed for compilation? I would appreciate your guidance.

**Reply by coryan** [reply]
> I also used the ldd command to inspect my quickstart, and I noticed that there is no SSL-related information.

That is because you compiled using vcpkg, which creates only static libraries by default.  Most recent versions of vcpkg use OpenSSL 3.x.  Why do you think your application is using OpenSSL 1.x?

> Is there anything wrong with the steps I followed for compilation?

The compilation steps look correct, and they produce a working `quickstart`.  I think we have ruled out obvious networking problems in the development environment. We need to examine other possibilities at this point.

Earlier you said:

  >  I encounter the following error:
  > "Error in non-idempotent operation SynthesizeSpeech: failed to connect to all addresses; last error: UNKNOWN: ipv6:[2404:6800:4004:823::200a]:443: Network is unreachable."

Did that error happen once, from time-to-time or persistently?  I have been assuming that the error happens more or less all the time.  Is that a valid assumption?

Did the error happen in the same environment as your quickstart?  Or does this error only happen when deployed to production or when not in the development environment?

I assume you are trying to use this into a larger application or program?  If so, any significant differences in how you build for this program vs. the quickstart?  For example: are you dynamically loading the code via some plugin?

**Reply by depinglin85** [reply]
Thank you very much for your response.

I would like to develop a FreeSWITCH module using gRPC Text-to-Speech. 
Currently, FreeSWITCH only supports OpenSSL 1.x, so I suspect the issue is due to OpenSSL version incompatibility. 
My development environment is the same as the one used for compiling the quickstart.

Could you please guide me on how to compile without using vcpkg on the Debian system? I only found vcpkg-based compilation instructions in the readme file.

**Reply by coryan** [reply, marked as answer]
> Currently, FreeSWITCH only supports OpenSSL 1.x, so I suspect the issue is due to OpenSSL version incompatibility.
> My development environment is the same as the one used for compiling the quickstart.

Okay.  I think the problem is using both OpenSSL 1.x and OpenSSL 3.x at the same time.  The vcpkg-based build uses OpenSSL 3.x, FreeSWITCH uses OpenSSL 1.x, and when FreeSWITCH loads the module (I reckon this is some kind of `.so` file) nothing works like it should.

> Could you please guide me on how to compile without using vcpkg on the Debian system? 

Certainly, look at:

https://github.com/googleapis/google-cloud-cpp/blob/main/doc/packaging.md

There should be step by step instructions for Debian 11 in that file.

**Reply by depinglin85** [reply]
Thank you very much
I have successfully compiled grpc and google-cloud-cpp, but I cannot find "google_cloud_cpp_texttospeech" in my own project's CMakeLists.txt.

`find_package(google_cloud_cpp_texttospeech REQUIRED)`
error:
`CMake Error at CMakeLists.txt:14 (find_package):
  By not providing "Findgoogle_cloud_cpp_texttospeech.cmake" in
  CMAKE_MODULE_PATH this project has asked CMake to find a package
  configuration file provided by "google_cloud_cpp_texttospeech", but CMake
  did not find one.`

Can you give me some advice? thanks

**Reply by coryan** [reply]
> Can you give me some advice?

Either the CMake configuration file is not installed or your CMake invocation is missing a search path (as the error suggests).

The documentation says:

```
# Pick a location to install the artifacts, e.g., `/usr/local` or `/opt`
PREFIX="${HOME}/google-cloud-cpp-installed"
```

First, verify the file is in that directory.  Something like:

```
find $PREFIX -name 'google_cloud_cpp_texttospeech*'
```

I would expect the following files to be found:

```
..../cmake/google_cloud_cpp_texttospeech/google_cloud_cpp_texttospeech-targets-noconfig.cmake
..../cmake/google_cloud_cpp_texttospeech/google_cloud_cpp_texttospeech-config.cmake
..../cmake/google_cloud_cpp_texttospeech/google_cloud_cpp_texttospeech-targets.cmake
..../cmake/google_cloud_cpp_texttospeech/google_cloud_cpp_texttospeech-config-version.cmake
```

If the files are not found, you probably missed one of the installation steps.  If the files are found, then your CMake configuration step  (for your application) is missing something like `-DCMAKE_PREFIX_PATH=${PREFIX}`.

**Reply by depinglin85** [reply]
Thank you very much
As you mentioned, I forgot to include -DGOOGLE_CLOUD_CPP_ENABLE=__ga_libraries__. 
With your help, I successfully compiled google-cloud-cpp and built the texttospeech quickstart, which is now running smoothly.
```
ldd quickstart 
	linux-vdso.so.1 (0x00007ffd363e6000)
	libsystemd.so.0 => /lib/x86_64-linux-gnu/libsystemd.so.0 (0x00007f8d6b608000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f8d6b5e6000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f8d6b419000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8d6b2d5000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f8d6b2bb000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8d6b0e7000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f8d6cc43000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f8d6b0db000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f8d6b0b3000)
	libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007f8d6afd8000)
	liblz4.so.1 => /lib/x86_64-linux-gnu/liblz4.so.1 (0x00007f8d6afb5000)
	libgcrypt.so.20 => /lib/x86_64-linux-gnu/libgcrypt.so.20 (0x00007f8d6ae95000)
	libgpg-error.so.0 => /lib/x86_64-linux-gnu/libgpg-error.so.0 (0x00007f8d6ae6d000)
```
I have successfully compiled my freeswitch module, but I encountered an error when loading the module.

`undefined symbol: _ZN6google5cloud12texttospeech2v120VoiceSelectionParamsC1EPNS_8protobuf5ArenaE**`
Can you give me some advice? thanks

**Reply by coryan** [reply]
> I have successfully compiled my freeswitch module, but I encountered an error when loading the module.
> 
> `undefined symbol: _ZN6google5cloud12texttospeech2v120VoiceSelectionParamsC1EPNS_8protobuf5ArenaE**`

Okay.

> Can you give me some advice?

Looks like a missing dependency.  You are linking `google-cloud-cpp::texttospeech` which *should* provide all the dependencies you need.  Maybe look into whether `mod_google_tts` has all the right code linked in.  Maybe run something like:


```shell
ldd libmod_google_tts.so
```

If the `libgoogle_cloud_cpp*` libraries are statically linked they will not appear in this command.  That would be my recommended approach, because modules like this are easier to manage if they have fewer run-time dependencies. It is not the *only* approach, just the one I would use if I was trying to do this.  In any case, if the libraries are statically linked, I would try to debug the link commands from your CMake file.  Use `--verbose` ([ref](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-build-v)) when compiling the code (some older versions of CMake use `-- VERBOSE=1` or something similar.

If the `libgoogle_cloud_cpp*` libraries are dynamically linked into `libmod_google_tts.so`, then make sure the right `LD_LIBRARY_PATH` (or the right `ld.so` configuration) is set at run-time.

**Reply by depinglin85** [reply]
Thank you for your assistance. I have successfully resolved all the issues.
Now I am able to make proper calls to the Text-to-Speech API within my module.

**Reply by coryan** [reply]
Glad to hear we were able to help.  If you have some time to describe what fixed the issue that would be useful to anybody running into  similar problems in the future.

**Comment by depinglin85** [comment]
Initially, I wanted to develop a Freeswitch mod using the Text-to-Speech API. I successfully compiled the Freeswitch mod, but encountered the following error when trying to load the mod.
`"Error in non-idempotent operation SynthesizeSpeech: failed to connect to all addresses; last error: UNKNOWN: ipv6:[2404:6800:4004:823::200a]:443: Network is unreachable."`

Later, I discovered that the Google Cloud API was compiled with OpenSSL 3.x (via vcpkg), while my Freeswitch was using OpenSSL 1.2. To resolve this, I recompiled the Google Cloud API using CMake and also recompiled my Freeswitch mod. However, I encountered the following error during the loading process.
https://github.com/googleapis/google-cloud-cpp/blob/main/doc/packaging.md
`undefined symbol: _ZN6google5cloud12texttospeech2v120VoiceSelectionParamsC1EPNS_8protobuf5ArenaE**`

Later, I realized that the issue was due to my build tool (sh) not properly removing vcpkg. Once I removed -DCMAKE_TOOLCHAIN_FILE, there were no more issues.
